packages/autorest.python/autorest/codegen/templates/model_base.py.jinja2

```diff
@@ -128,10 +128,24 @@ def _is_readonly(p):
 class AzureJSONEncoder(JSONEncoder):
     """A JSON encoder that's capable of serializing datetime objects and bytes."""
 
+    def __init__(self, *args, exclude_readonly: bool = False, exclude_none: bool = False, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.exclude_readonly = exclude_readonly
+        self.exclude_none = exclude_none
+
     def default(self, o):  # pylint: disable=too-many-return-statements
         if _is_model(o):
-            readonly_props = [p._rest_name for p in o._attr_to_rest_field.values() if _is_readonly(p)]
-            return {k: v for k, v in o.items() if k not in readonly_props}
+            result = dict(o.items())
+            if self.exclude_readonly:
+                readonly_props = [p._rest_name for p in o._attr_to_rest_field.values() if _is_readonly(p)]
+                for k in readonly_props:
+                    if k in result:
+                        result.pop(k)
+            if self.exclude_none:
+                for k in list(result.keys()):
+                    if result[k] is None or isinstance(result[k], _Null):
```

Comment:
we don't want to pop `_Null`, only `None`

Reply:
as an aside, we should never deserialize to `azure.core.serialization.NULL` as well, this is an input only concept.

There isn't clear information on when we should deserialize to that. Our current behavior in our released SDKs is to ignore `None` values on deserialization (which is what we also do on input). Therefore, it would be breaking to deserialization a `None` to `azure.core.serialization.NULL`, so we should continue ignoring `None` in response bodies, and only send `null` to the server for input if someone explicitly used the sentinel `azure.core.serialization.NULL`